### DISCUSSION TITLE
I just realized that uuid for thes same photo is different on different computers.

### DISCUSSION THREAD

**Discussion by chengguangnan** [RhetTbull/osxphotos#1386] [General]
The uique uuid is `cloudassetid` /  `fingerprint`.

**Comment by RhetTbull** [comment]
Yes. UUID is unique only to the specific library. If you repair the library or create new library and re-download from iCloud, then UUID will be different. If you move a backup between computers, you can use the `osxphotos exportdb --migrate-photos-library` command to update the UUIDs in the export database.

osxphotos doesn't use `cloudassetid` because these exist only for photos in an iCloud database. It also doesn't use fingerprint because Photos computes fingerprints only for photos, not for videos, and because fingerprint is unique only to the actual file not the asset in the database. That is, you could have duplicates in your library that have different metadata, are in different albums, etc. but they have the same fingerprint.

**Reply by chengguangnan** [reply]
Thank you. Very informative!

**Reply by RhetTbull** [reply]
By the way, the logic used for the `--migrate-photos-library` command is this:

If cloud GUID exists, use that because it's guaranteed to be the same asset
If not, and fingerprint exists, use `filename:fingerprint`
If fingerprint does not exist, use `filename:filesize`

If photo is in a legacy "shared album", none of these exist for the photo so the "fingerprint" used is "cloud_owner_hashed_id:heigh:width:isphoto:isvideo:date"

This isn't perfect and there could be false matches but it is not likely and it's the best algorithm I've been able to come up with. If someone can devise a more robust matching algorithm I'm happy to accept a PR! 

This logic is important and is used in other places in the code -- for example, for the `sync` command which syncs metadata between two libraries and the `import` command which checks for duplicates when importing. 

I'm not sure why Photos doesn't compute fingerprint for videos -- would be more useful if it did. The fingerprint method (which is a private undocumented method in the Photos framework which I've discovered and figured out how to call) works fine on video files. My guess is that because videos can be very large, checking the fingerprint on import could be slow and thus it's avoided. On modern Macs though the algorithm runs very fast. It computes some sort of hash though I've not figured out which exactly (but also not relevant as osxphotos needs to match Photos).

**Reply by gwood** [reply]
Have you tried xxHash for the hash to see if the time to run and/or result matches the output you are getting with Apple private undocumented method? The GitHub README for the code also list lots of other hash methods that are fast.

https://github.com/Cyan4973/xxHash

**Reply by RhetTbull** [reply]
I've not tried. Will take a look when I get a chance. Unless the hash matches exactly what is produced by Apple's fingerprint method, it won't be useful for OSXPhotos even if it's very fast.
